### ISSUE TITLE
Task dropped in wrong thread when aborting LocalSet task

### ISSUE THREAD

**Issue by Darksonn** [tokio-rs/tokio#3929]
When aborting a task with `JoinHandle::abort`, the future is dropped in the thread calling `abort` if the task is not currently being executed. This is incorrect for tasks spawned on a `LocalSet`. See this example that exploits it to send a non-send value to a different thread:
```Rust
use tokio::task::LocalSet;
use std::time::Duration;

struct PrintOnDrop {
    created_on: std::thread::ThreadId,
    marker: std::marker::PhantomData<*const ()>,
}

impl PrintOnDrop {
    fn new() -> Self {
        Self {
            created_on: std::thread::current().id(),
            marker: std::marker::PhantomData,
        }
    }
}
impl Drop for PrintOnDrop {
    fn drop(&mut self) {
        if std::thread::current().id() != self.created_on {
            println!("non-Send value dropped in another thread!");
        }
    }
}

#[tokio::main]
async fn main() {
    let ls = LocalSet::new();
    let pod = PrintOnDrop::new();
    let jh = ls.spawn_local(async move {
        std::future::pending::<()>().await;
        drop(pod);
    });
    
    std::thread::spawn(move || {
        std::thread::sleep(Duration::from_millis(10));
        jh.abort();
    });
    
    ls.await;
}
```

**Comment by Darksonn** [comment]
Note: I have some follow-up PRs to #3909 planned, which will fix this issue.

**Comment by Darksonn** [comment]
This was fixed in #3934. The fix has been backported to all minor releases that are at most 3 months old which includes `1.5` and newer.

**Comment by Shnatsel** [comment]
To double-check, this is the list of all versions published to crates.io, and whether they're vulnerable or not according to the current version specification:

```
0.0.0 OK
0.1.0 OK
0.1.1 OK
0.1.2 OK
0.1.3 OK
0.1.4 OK
0.1.5 OK
0.1.6 OK
0.1.7 OK
0.1.8 OK
0.1.9 OK
0.1.10 OK
0.1.11 OK
0.1.12 OK
0.1.13 OK
0.1.14 OK
0.1.15 OK
0.1.16 OK
0.1.17 OK
0.1.18 OK
0.1.19 OK
0.1.20 OK
0.1.21 OK
0.1.22 OK
0.2.0-alpha.1 OK
0.2.0-alpha.2 OK
0.2.0-alpha.3 OK
0.2.0-alpha.4 OK
0.2.0-alpha.5 OK
0.2.0-alpha.6 OK
0.2.0 OK
0.2.1 OK
0.2.2 OK
0.2.3 OK
0.2.4 OK
0.2.5 OK
0.2.6 OK
0.2.7 OK
0.2.8 OK
0.2.9 OK
0.2.10 OK
0.2.11 OK
0.2.12 OK
0.2.13 OK
0.2.14 OK
0.2.15 OK
0.2.16 OK
0.2.17 OK
0.2.18 OK
0.2.19 OK
0.2.20 OK
0.2.21 OK
0.2.22 OK
0.3.0 vulnerable
0.3.1 vulnerable
0.3.2 vulnerable
0.3.3 vulnerable
0.2.23 OK
0.3.4 vulnerable
0.3.5 vulnerable
0.2.24 OK
0.3.6 vulnerable
1.0.0 vulnerable
1.0.1 vulnerable
1.0.2 vulnerable
1.1.0 vulnerable
0.2.25 OK
0.3.7 vulnerable
1.0.3 vulnerable
1.1.1 vulnerable
1.2.0 vulnerable
1.3.0 vulnerable
1.4.0 vulnerable
1.5.0 vulnerable
1.6.0 vulnerable
1.6.1 vulnerable
1.6.2 vulnerable
1.7.0 vulnerable
1.7.1 vulnerable
1.8.0 vulnerable
1.5.1 OK
1.6.3 OK
1.7.2 OK
1.8.1 OK
```

Anyway, my apologies for the inconvenience, and thanks for reporting the issue!
